logger.list_entries optimization using equals  and docs update

waprin opened the pull request:
Two changes here:

logger.list_entries() was using just the logger name rather than the fully qualified name, you can see an example on [the docs ](https://cloud.google.com/logging/docs/view/advanced_filters#names)that from v1 to v2 the change from log to logName also changed to expect fully qualified name.

The query was also using the colon operator, which is a "has" operator. Problem 1 is that this is a slower query because it's not indexed, problem 2 is that I'm pretty sure there's a bug because it's not working.

Second set of changes is the docs, just some random small discrepancies between them and the actual API. The biggest discrepancy was in the Sink constructor, the samples all use keyword arguments and special names for different sink types, the actual code seems to take these all as general positional arguments.

dhermes commented:
I suppose I should let @tseaver take the lead here

daspecster commented on docs/logging-usage.rst:
```diff
@@ -250,10 +250,11 @@ Create a Cloud Pub/Sub sink:
    ...     "robots-pubsub",
-   ...     filter='log:apache-access AND textPayload:robot')
-   >>> sink.pubsub_topic = 'projects/my-project/topics/my-topic'
+   ...      'log:apache-access AND textPayload:robot', 
```
Small note, the spacing here seems to be a little different from the other examples.
Also, there's a space at the end of the line as well.

waprin replied:
deleted space at end, I'm not sure what specifically you mean by it being different. I just manually entered spaces to align it with the lines above.

tseaver commented on gcloud/logging/logger.py:
```diff
@@ -298,7 +298,7 @@ def list_entries(self, projects=None, filter_=None, order_by=None,
                   value as ``page_token``).
         """
-        log_filter = 'logName:%s' % (self.name,)
+        log_filter = 'logName=%s' % (self.full_name,)
```
Do I understand correctly: the original version "worked" because the two errors (substring operator + short name) cancelled out?

waprin replied:
No, the original version doesn't work. I'm guessing the API changed underneath it. But logger.list_entries() does not work for me as is (would be curious if you tried now whether it works for you).

The original version actually should work since the fully qualified name should contain its own name and the colon operator is contains. However, = is better anyway since it's indexed, and the colon operator currently seems to be buggy.

I have been playing with the API and discovered the following things:

projectId can take either just a projectId or a list of projectIds. The docs say it should take a list.

If you don't give it a list, both = and : operator will work, but : operator will return an extraneous token, that if used will keep returning just itself.

If you do give it a list ,only = will work, : operator will return 0 results with a nextPageToken that keeps returning itself.

I am following up internally about these issues but I think even if these bugs are fixed, this is the correct change to make since it's indexed.

waprin replied:
Ok, now I see this in fact not a bug at all on the client side or the server side.The list syntax difference was just my mistake. Without a resource type and without an equal sign to use an index, the query is just extremely slow. The same nextPageToken keeps coming back because it's indicating that the same query is still taking place, even if there are still no results. The code will work as is, but can take several minutes to get results even on a brand new project.

So yeah, we should still make this change, but it's an optimization, not a bug fix. Given how sluggish the logging API can be, we might even consider changing all the samples to polling the request until the token is None.

tseaver commented on docs/logging-usage.rst:
```diff
@@ -220,8 +220,8 @@ Create a Cloud Storage sink:
-   ...     filter='log:apache-access AND textPayload:robot')
-   >>> sink.storage_bucket = "my-bucket-name"
+   ...     'log:apache-access AND textPayload:robot',
+   ...     "my-bucket-name")
```
The `destination` argument actually needs to be a URI, in one of the three forms corresponding to the three services.  In this case, it should be `storage.googleapis.com/my-bucket-name`.

waprin replied:
done

tseaver commented on docs/logging-usage.rst:
```diff
@@ -236,8 +236,8 @@ Create a BigQuery sink:
-   ...     filter='log:apache-access AND textPayload:robot')
-   >>> sink.bigquery_dataset = "projects/my-project/datasets/my-dataset"
+   ...     'log:apache-access AND textPayload:robot',
+   ...     "projects/my-project/datasets/my-dataset")
```
Should be `bigquery.googleapis.com/projects/projects/my-project/datasets/my-dataset`.

waprin replied:
done

tseaver commented on docs/logging-usage.rst:
```diff
@@ -250,10 +250,11 @@ Create a Cloud Pub/Sub sink:
-   ...     filter='log:apache-access AND textPayload:robot')
-   >>> sink.pubsub_topic = 'projects/my-project/topics/my-topic'
+   ...      'log:apache-access AND textPayload:robot', 
+   ...      'projects/my-project/topics/my-topic')
```
Should be `pubsub.googleapis.com/projects/my-project/topics/my-topic`.

waprin replied:
done

waprin commented:
Ok, addressed all review comments, changed commit messages/PR name to reflect that logging changes are optimizations rather then bugfixes,assuming Travis passes it's ready for another pass.

tseaver commented:
@waprin Thanks for the research, as well as the patch!
